### PR TITLE
perf(push, findMergeBase): improve find merge base speed

### LINKED ISSUE #1131: `findMergeBase` walks oids multiple times
Hey William and isomorphic-git team :wink:

We are going through a major update on our CLI at Clever Cloud (0.37.0 to latest) and we encountered some perf problems.

The first one concerns `findMergeBase`.

## Test context

We ran our tests on 2 projects:

Thank you `git-sizer`

* repo WWW
  * 2.2k commits
  * 10.7k trees
  * 164k tree entries
  * 4.7k blobs
  * 4.7 MiB for biggest blob
* repo CONSOLE
  * 3.4k commits
  * 14.8k trees
  * 266k tree entries
  * 8.1k blobs
  * 443 KiB for biggest blob

We timed the execution of a `push` in two situations:

* pushing latest master on empty remote repo
* pushing latest master on repo with only the first commit

## Tests results

* On the WWW repo, the process took more than 25min (we stopped it manually).
* We added timers around main steps of `push` and it seems like `findMergeBase` is "never" ending.

In an empty remote repo scenario, `findMergeBase` is straight forward so we though we could just early return `[]` when `oids` contain `'0000000000000000000000000000000000000000'`.

This worked great but when we tried to push the whole repo on a remote that already had just the first commit, we still had the "very slow" problem.

* We digged a bit more on how `findMergeBase` works.
* We logged how many loops we had and what was inside `heads`.
* For our WWW repo (2.2k commits) we reached more 5k loops after only 30 seconds...
* We also noticed that `heads` contained lots of duplicates `oid`+`index`.
  * We observed a `heads` with 768 `{oid,index}` objects but 4 unique entries the other 764 were duplicates.

## Proposed solution

We think it's what causing the slow behaviour in our two situations so we patched `findMergeBase` so it creates a `newheads` with only unique values using a `Map` keyed by `oid:index`.

:tada: This was a major improvement:
* Project WWW had a `findMergeBase` timed at ~1 sec
* Project CONSOLE had a `findMergeBase` timed at <1 sec

### PR DESCRIPTION
**Pull request by hsablonniere**
This is a proposition to fix #1131

### CODE REVIEW

File: __tests__/test-findMergeBase.js [hunk 1]
```diff
@@ -5,7 +5,7 @@ const { makeFixture } = require('./__helpers__/FixtureFS.js')
 
 // These have been checked with
 // GIT_DIR=__tests__/__fixtures__/test-findMergeBase.git git merge-base -a --octopus COMMITS
-describe('findMergeBase', () => {
+describe.only('findMergeBase', () => {
```
Comment (wmhilton) [inline __tests__/test-findMergeBase.js:8]:
```suggestion
describe('findMergeBase', () => {
```

**Comment by wmhilton** [comment]
I'm going to move the '0000000000000000000000000000000000000000' check up a level to the `push` command, because '0000000000000000000000000000000000000000' isn't a "real" commit hash... it's really just used in the wire protocol as a placeholder in branch creation/deletion.

**Comment by karma-pr-reporter** [comment]
## Test Results for 2950f6d0c954fe0452174064a4447b5b10a673c8:
| Browser                                 | Pass | Skip | Fail | Time   | Disconnect? |
| --------------------------------------- | ---- | ---- | ---- | ------ | ----------- |
| Chrome 79.0.3945 (Windows 10 0.0.0)     | 363  | 0    | 0    | 3m:25s | false       |
| Chrome Mobile 80.0.3987 (Android 0.0.0) | 363  | 0    | 0    | 1m:55s | false       |
| Firefox 76.0.0 (Ubuntu 0.0.0)           | 363  | 0    | 0    | 0m:53s | false       |
| HeadlessChrome 0.0.0 (Linux 0.0.0)      | 363  | 0    | 0    | 0m:46s | false       |
| Mobile Safari 13.0.0 (iOS 13.0.0)       | 363  | 0    | 0    | 2m:51s | false       |
| Safari 13.0.3 (Mac OS X 10.15.1)        | 363  | 0    | 0    | 3m:29s | false       |

### Longest running tests
- 11.498 secs - clone clone with noTags - _Chrome_
- 9.153 secs - clone clone with noTags - _Firefox_
- 8.512 secs - fetch fetch --prune-tags from git-http-mock-server - _Chrome_
- 8.17 secs - clone clone with noTags - _Mobile Safari_
- 7.716 secs - clone clone with noTags - _HeadlessChrome_
- 7.325 secs - walk can populate type, mode, oid, and content - _Chrome_
- 7.163 secs - clone clone with noTags - _Safari_
- 7.121 secs - fetch fetch --prune-tags from git-http-mock-server - _Mobile Safari_
- 6.597 secs - statusMatrix statusMatrix with filepaths - _Chrome_
- 6.21 secs - clone clone with noTags - _Chrome Mobile_
- 5.969 secs - fetch fetch --prune from git-http-mock-server - _Safari_

**Comment by wmhilton** [comment]
Huh. This is kind of interesting. I saw you used `Array.from(newheads.values())` whereas a couple of lines earlier I'd used `[...result]`. Generally speaking, I like spread syntax. But that got me thinking... is there a performance difference?

Turns out there is and it's counter-intuitive! Chrome on the left, Firefox on the right:
image

Array.from is 4x _faster_ in Chrome, but about 20% _slower_ in Firefox. Riddle me that lol.
https://esbench.com/bench/5edbd14e12464000a01e46c3

Anyhoo, tests pass now yay. Just a couple final things to tweak, then I'll get this merged.

**Comment by wmhilton** [comment]
Thank you for your patience! ❤️

**Pull request merged**